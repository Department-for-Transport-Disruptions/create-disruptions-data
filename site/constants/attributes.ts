export const CD_SUMMARY = "summary";

export const CD_DESCRIPTION = "description";

export const CD_DISRUPTION_TYPE = "disruptionType";

export const CD_ASSOCIATED_LINK = "associatedLink";

export const CD_DISRUPTION_REASON = "disruptionReason";

export const CD_DISRUPTION_REAPEATS = "disruptionRepeats";

export const CD_NO_VALIDITY_END_DATE_TIME = "validityNoEndDateTime";

export const CD_NO_PUBLISH_END_DATE_TIME = "publishNoEndDateTime";

export const CD_VALIDITY_START_DATE_DAY = "validityStartDateDay";

export const CD_VALIDITY_START_DATE_MONTH = "validityStartDateMonth";

export const CD_VALIDITY_START_DATE_YEAR = "validityStartDateYear";

export const CD_VALIDITY_START_HOUR = "validityStartHour";

export const CD_VALIDITY_START_MINUTES = "validityStartMinute";

export const CD_VALIDITY_END_DATE_DAY = "validityEndDateDay";

export const CD_VALIDITY_END_DATE_MONTH = "validityEndDateMonth";

export const CD_VALIDITY_END_DATE_YEAR = "validityEndDateYear";

export const CD_VALIDITY_END_HOUR = "validityEndHour";

export const CD_VALIDITY_END_MINUTES = "validityEndMinute";

export const CD_PUBLISH_START_DATE_DAY = "publishStartDateDay";

export const CD_PUBLISH_START_DATE_MONTH = "publishStartDateMonth";

export const CD_PUBLISH_START_DATE_YEAR = "publishStartDateYear";

export const CD_PUBLISH_START_HOUR = "publishStartHour";

export const CD_PUBLISH_START_MINUTES = "publishStartMinute";

export const CD_PUBLISH_END_DATE_DAY = "publishEndDateDay";

export const CD_PUBLISH_END_DATE_MONTH = "publishEndDateMonth";

export const CD_PUBLISH_END_DATE_YEAR = "publishEndDateYear";

export const CD_PUBLISH_END_HOUR = "publishEndHour";

export const CD_PUBLISH_END_MINUTES = "publishEndMinute";
