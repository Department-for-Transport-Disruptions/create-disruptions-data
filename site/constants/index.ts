import {
    MiscellaneousReason,
    PersonnelReason,
    EnvironmentReason,
    EquipmentReason,
} from "@create-disruptions-data/shared-ts/enums";

export const SUPPORT_EMAIL_ADDRESS = "bodshelpdesk@kpmg.co.uk";

export const SERVICE_EMAIL_ADDRESS = "bodshelpdesk@kpmg.co.uk";

export const SUPPORT_PHONE_NUMBER = "0800 123 1234";

export const DISRUPTION_REASONS: {
    value: MiscellaneousReason | PersonnelReason | EnvironmentReason | EquipmentReason;
    display: string;
}[] = [
    {
        value: MiscellaneousReason.roadWorks,
        display: "Road Works",
    },
    {
        value: MiscellaneousReason.vandalism,
        display: "Vandalism",
    },
    {
        value: MiscellaneousReason.routeDiversion,
        display: "Route Diversion",
    },
    {
        value: MiscellaneousReason.specialEvent,
        display: "Special Event",
    },
];

export const DISRUPTION_TYPES = ["planned", "unplanned"];

export const CD_DATE_FORMAT = "DD/MM/YYYY";
export const CONTACT_FEEDBACK_QUESTION = "Did you contact us for assistance at any point?";
export const SOLVE_FEEDBACK_QUESTION = "Did we solve your problem?";
export const HEAR_ABOUT_US_FEEDBACK_QUESTION = "How did you hear about our service?";
export const GENERAL_FEEDBACK_QUESTION = "Please let us know any feedback or suggestions for improvements you may have";

export const COOKIES_POLICY_COOKIE = "cdd-cookies-policy";

export const COOKIE_PREFERENCES_COOKIE = "cdd-cookie-preferences-set";

export const oneYearInSeconds = 31556952;

export const ID_TOKEN_COOKIE = "cdd-id-token";

export const COOKIES_DISRUPTION_INFO = "disruption-info";

export const COOKIES_DISRUPTION_ERRORS = "disruption-errors";

export const COOKIES_ADD_CONSEQUENCE_INFO = "cdd-consequence-info";

export const COOKIES_ADD_CONSEQUENCE_ERRORS = "cdd-consequence-errors";

export const CREATE_DISRUPTION_PAGE_PATH = "/create-disruption";

export const ERROR_PATH = "/_error";

export const ADD_CONSEQUENCE_PAGE_PATH = "/type-of-consequence";

export const STAGE = process.env.STAGE || "dev";
