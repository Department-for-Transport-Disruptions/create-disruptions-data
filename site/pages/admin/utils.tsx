import resendInvite from "../api/admin/resend-invite.api";

const createUserActionsLink = (
    key: string,
    index: number,
    username: string,
    userGroup: string,
    showResendInvite?: boolean,
) => {
    return (
        <>
            {showResendInvite ? (
                <>
                    <button
                        key={`${key}${index ? `-${index}` : ""}`}
                        className="govuk-link"
                        onClick={() => resendInvite(username, userGroup)}
                    >
                        Resend invite
                    </button>
                    <br />
                    <button
                        key={`${key}${index ? `-remove-${index}` : "-remove"}`}
                        className="govuk-link"
                        onClick={() => removeUser(username)}
                    >
                        Remove
                    </button>
                </>
            ) : (
                <button
                    key={`${key}${index ? `-${index}` : ""}`}
                    className="govuk-link"
                    onClick={() => removeUser(username)}
                >
                    Remove
                </button>
            )}
        </>
    );
};
