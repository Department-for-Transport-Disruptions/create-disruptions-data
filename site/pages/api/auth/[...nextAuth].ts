import NextAuth from "next-auth";
import CognitoProvider from "next-auth/providers/cognito";

export default NextAuth({
    providers: [
        CognitoProvider({
            clientId: "",
            clientSecret: "",
            issuer: "",
        }),
    ],
    pages: {
        signIn: "",
        newUser: "",
    },
    session: {
        strategy: "jwt",
    },
});
