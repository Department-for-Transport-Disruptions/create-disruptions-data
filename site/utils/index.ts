import { ErrorInfo } from "interfaces";
import { ServerResponse } from "http";
import { NextApiResponse } from "next";

export const buildTitle = (errors: ErrorInfo[], title: string): string => {
    if (errors.length > 0) {
        return `Error: ${title}`;
    }

    return title;
};

export const redirectTo = (res: NextApiResponse | ServerResponse, location: string): void => {
    res.writeHead(302, {
        Location: location,
    });
    res.end();
};
