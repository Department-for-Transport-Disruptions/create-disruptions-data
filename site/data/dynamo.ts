import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";
import { ptSituationElementSchema } from "@create-disruptions-data/shared-ts/siriTypes.zod";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({ region: "eu-west-2" }));

const tableName = process.env.TABLE_NAME as string;

export const getDisruptionsDataFromDynamo = async () => {
    logger.info("Getting disruptions data from DynamoDB table...");

    const dbData = await ddbDocClient.send(
        new QueryCommand({
            TableName: tableName,
            KeyConditionExpression: "PK = :i",
            ExpressionAttributeValues: {
                ":i": "1",
            },
        }),
    );

    return dbData.Items?.map((item) => {
        const parsedItem = ptSituationElementSchema.safeParse(item);

        if (!parsedItem.success) {
            logger.error("Error parsing disruption from dynamo");
            logger.error(parsedItem.error.stack);
            return null;
        }

        return parsedItem.data;
    }).filter(notEmpty);
};
