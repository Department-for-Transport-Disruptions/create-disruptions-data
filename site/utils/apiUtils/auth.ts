import { decodeJwt } from "jose";
import { NextApiRequest } from "next";
import { parseCookies } from "nookies";
import { z } from "zod";
import { IncomingMessage } from "http";
import { COOKIES_ID_TOKEN } from "../../constants";
import { getOrganisationInfoById } from "../../data/dynamo";

const sessionSchema = z
    .object({
        sub: z.string().uuid(),
        email: z.string().email(),
        "custom:orgId": z.string().uuid(),
        "cognito:groups": z.array(
            z.union([
                z.literal("system-admins"),
                z.literal("org-admins"),
                z.literal("org-publishers"),
                z.literal("org-staff"),
            ]),
        ),
    })
    .transform((item) => {
        const isSystemAdmin = item["cognito:groups"].includes("system-admins");
        const isOrgAdmin = isSystemAdmin || item["cognito:groups"].includes("org-admins");
        const isOrgPublisher = isSystemAdmin || isOrgAdmin || item["cognito:groups"].includes("org-publishers");
        const isOrgStaff =
            isSystemAdmin || isOrgAdmin || isOrgPublisher || item["cognito:groups"].includes("org-staff");

        return {
            username: item.sub,
            email: item.email,
            orgId: item["custom:orgId"],
            isSystemAdmin,
            isOrgAdmin,
            isOrgPublisher,
            isOrgStaff,
        };
    });

const sessionSchemaWithOrgDetail = sessionSchema.transform(async (item) => {
    const orgDetail = await getOrganisationInfoById(item.orgId);

    return {
        ...item,
        orgName: orgDetail?.name ?? "",
        adminAreaCodes: orgDetail?.adminAreaCodes ?? [],
    };
});

export type Session = z.infer<typeof sessionSchema>;
export type SessionWithOrgDetail = z.infer<typeof sessionSchemaWithOrgDetail>;

export const getSession = (req: NextApiRequest | IncomingMessage): Session | null => {
    const cookies = parseCookies({ req });
    const idToken = cookies[COOKIES_ID_TOKEN];

    if (idToken) {
        return sessionSchema.parse(decodeJwt(idToken));
    }

    return null;
};

export const getSessionWithOrgDetail = async (
    req: NextApiRequest | IncomingMessage,
): Promise<SessionWithOrgDetail | null> => {
    const cookies = parseCookies({ req });
    const idToken = cookies[COOKIES_ID_TOKEN];

    if (idToken) {
        return sessionSchemaWithOrgDetail.parseAsync(decodeJwt(idToken));
    }

    return null;
};
