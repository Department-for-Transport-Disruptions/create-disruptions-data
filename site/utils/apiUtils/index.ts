import { Consequence } from "@create-disruptions-data/shared-ts/disruptionTypes";
import { MAX_CONSEQUENCES } from "@create-disruptions-data/shared-ts/disruptionTypes.zod";
import { SocialMediaPostStatus } from "@create-disruptions-data/shared-ts/enums";
import { NextApiRequest, NextApiResponse } from "next";
import { parseCookies, setCookie } from "nookies";
import { TwitterApi } from "twitter-api-v2";
import { z } from "zod";
import { IncomingMessage, ServerResponse } from "http";
import { notEmpty } from "..";
import {
    COOKIES_POLICY_COOKIE,
    COOKIE_CSRF,
    COOKIES_ID_TOKEN,
    COOKIE_PREFERENCES_COOKIE,
    DISRUPTION_DETAIL_PAGE_PATH,
    REVIEW_DISRUPTION_PAGE_PATH,
    COOKIES_REFRESH_TOKEN,
} from "../../constants";
import { upsertConsequence } from "../../data/dynamo";
import { getAccessToken, publishToHootsuite } from "../../data/hootsuite";
import { getAuthedTwitterClient, sendTweet } from "../../data/twitter";
import { TooManyConsequencesError } from "../../errors";
import { PageState } from "../../interfaces";
import { SocialMediaPost } from "../../schemas/social-media.schema";
import logger from "../logger";

export const setCookieOnResponseObject = (
    cookieName: string,
    cookieValue: string,
    res: NextApiResponse | ServerResponse<IncomingMessage>,
    lifetime?: number,
    httpOnly = true,
    sameSite: "lax" | "strict" | "none" = "lax",
): void => {
    // From docs: All cookies are httponly by default, and cookies sent over SSL are secure by
    // default. An error will be thrown if you try to send secure cookies over an insecure socket.
    setCookie({ res }, cookieName, cookieValue, {
        path: "/",
        sameSite,
        secure: process.env.NODE_ENV !== "development",
        maxAge: lifetime,
        httpOnly,
    });
};

export const destroyCookieOnResponseObject = (
    cookieName: string,
    res: NextApiResponse | ServerResponse<IncomingMessage>,
): void => {
    setCookieOnResponseObject(cookieName, "", res, 0);
};

export const redirectTo = (res: NextApiResponse | ServerResponse, location: string): void => {
    res.writeHead(302, {
        Location: location,
    });
    res.end();
};

export const redirectToError = (
    res: NextApiResponse | ServerResponse,
    message?: string,
    context?: string,
    error?: Error,
): void => {
    if (message && context && error) {
        logger.error(message, { context, error: error.stack });
    }

    redirectTo(res, "/500");
};

export const cleardownCookies = (req: NextApiRequest, res: NextApiResponse) => {
    const cookies = parseCookies({ req });

    const saveList = [
        COOKIES_POLICY_COOKIE,
        COOKIE_PREFERENCES_COOKIE,
        COOKIES_ID_TOKEN,
        COOKIES_REFRESH_TOKEN,
        COOKIE_CSRF,
    ];

    Object.keys(cookies).forEach((cookie) => {
        if (!saveList.includes(cookie)) {
            destroyCookieOnResponseObject(cookie, res);
        }
    });
};

export const getPageState = <T>(errorCookie: string, schemaObject: z.ZodType<T>, disruptionId?: string, data?: T) => {
    const inputsProps: PageState<Partial<T>> = {
        errors: [],
        inputs: {},
        disruptionId: disruptionId || "",
    };

    if (errorCookie) {
        return { ...(JSON.parse(errorCookie) as PageState<Partial<T>>), disruptionId: disruptionId || "" };
    }

    if (disruptionId) {
        const parsedData = schemaObject.safeParse(data);

        if (parsedData.success) {
            inputsProps.inputs = parsedData.data;
        }
    }

    return inputsProps;
};

export const getReturnPage = (req: NextApiRequest) => {
    const queryParam = req.headers.referer?.split("?")[1];
    const decodedQueryParam = queryParam ? decodeURIComponent(queryParam) : null;
    return decodedQueryParam?.includes(REVIEW_DISRUPTION_PAGE_PATH) ||
        decodedQueryParam?.includes(DISRUPTION_DETAIL_PAGE_PATH)
        ? queryParam
        : null;
};

export const isDisruptionFromTemplate = (req: NextApiRequest) => {
    const queryParam = req.headers.referer?.split("?")[1];

    const decodedQueryParam = queryParam ? decodeURIComponent(queryParam) : null;
    return decodedQueryParam?.includes(DISRUPTION_DETAIL_PAGE_PATH) && decodedQueryParam.includes("template=true")
        ? queryParam
        : null;
};

export const delay = (ms: number) => new Promise((res) => setTimeout(res, ms));

export const publishSocialMedia = async (
    socialMediaPosts: SocialMediaPost[],
    orgId: string,
    isUserStaff: boolean,
    canPublish: boolean,
) => {
    const authedTwitterClients: Record<string, TwitterApi> = {};
    const hootsuiteAccessTokens: Record<string, string> = {};

    const uniqueTwitterSocialAccounts = new Set(
        socialMediaPosts.filter((post) => post.accountType === "Twitter").map((post) => post.socialAccount),
    );

    const uniqueHootsuiteSocialAccounts = new Set(
        socialMediaPosts.filter((post) => post.accountType === "Hootsuite").map((post) => post.socialAccount),
    );

    for (const socialAccount of uniqueHootsuiteSocialAccounts) {
        const accessToken = await getAccessToken(orgId, socialAccount);
        hootsuiteAccessTokens[socialAccount] = accessToken;
    }

    for (const socialAccount of uniqueTwitterSocialAccounts) {
        const authedClient = await getAuthedTwitterClient(orgId, socialAccount);

        if (authedClient) {
            authedTwitterClients[socialAccount] = authedClient;
        }
    }

    const socialMediaPromises = socialMediaPosts.map((post) => {
        if (post.status === SocialMediaPostStatus.pending) {
            if (post.accountType === "Twitter") {
                return sendTweet(orgId, post, isUserStaff, canPublish, authedTwitterClients[post.socialAccount]);
            }

            return publishToHootsuite(post, orgId, isUserStaff, canPublish, hootsuiteAccessTokens[post.socialAccount]);
        }

        return null;
    });

    await Promise.all(socialMediaPromises);
};

export const redirectToWithQueryParams = (
    req: NextApiRequest,
    res: NextApiResponse,
    queryParamsToForward: string[],
    location: string,
    paramsToAdd?: string[],
) => {
    const queryStringParams = queryParamsToForward
        .map((p) => {
            const paramValue = req.query[p];

            return paramValue ? `${p}=${paramValue.toString()}` : null;
        })
        .filter(notEmpty);

    redirectTo(
        res,
        `${location}${
            [...queryStringParams, ...(paramsToAdd || [])].length > 0
                ? `?${[...queryStringParams, ...(paramsToAdd || [])].join("&")}`
                : ""
        }`,
    );
};

export const handleUpsertConsequence = async (
    consequence: Consequence | Pick<Consequence, "disruptionId" | "consequenceIndex">,
    orgId: string,
    isOrgStaff: boolean,
    isTemplate: boolean,
    inputs: unknown,
    errorCookie: string,
    res: NextApiResponse,
) => {
    try {
        return await upsertConsequence(consequence, orgId, isOrgStaff, isTemplate);
    } catch (e) {
        if (e instanceof TooManyConsequencesError) {
            setCookieOnResponseObject(
                errorCookie,
                JSON.stringify({
                    inputs,
                    errors: [
                        {
                            id: "",
                            errorMessage: `Max consequence limit of ${MAX_CONSEQUENCES} has been reached`,
                        },
                    ],
                }),
                res,
            );
        }

        throw e;
    }
};
