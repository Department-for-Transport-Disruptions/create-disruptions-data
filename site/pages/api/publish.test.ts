import { PublishStatus, SocialMediaPostStatus } from "@create-disruptions-data/shared-ts/enums";
import MockDate from "mockdate";
import { afterAll, afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { DASHBOARD_PAGE_PATH, ERROR_PATH, REVIEW_DISRUPTION_PAGE_PATH } from "../../constants";
import * as dynamo from "../../data/dynamo";
import { FullDisruption } from "../../schemas/disruption.schema";
import { Organisation, defaultModes } from "../../schemas/organisation.schema";
import {
    DEFAULT_ORG_ID,
    disruptionWithConsequences,
    disruptionWithConsequencesAndSocialMediaPosts,
    disruptionWithNoConsequences,
    getMockRequestAndResponse,
    mockSession,
} from "../../testData/mockData";
import * as apiUtils from "../../utils/apiUtils";
import * as session from "../../utils/apiUtils/auth";
import * as disruptionApprovalEmailer from "../../utils/apiUtils/disruptionApprovalEmailer";
import publish from "./publish.api";

const defaultDisruptionId = "acde070d-8c4c-4f0d-9d8a-162843c10333";

const orgInfo: Organisation = {
    name: "DepartmentForTransport",
    adminAreaCodes: ["001", "002"],
    mode: defaultModes,
};

describe("publish", () => {
    const writeHeadMock = vi.fn();
    vi.mock("../../utils/apiUtils", async () => ({
        ...(await vi.importActual<object>("../../utils/apiUtils")),
        setCookieOnResponseObject: vi.fn(),
        destroyCookieOnResponseObject: vi.fn(),
        cleardownCookies: vi.fn(),
        publishToHootsuite: vi.fn(),
    }));

    vi.mock("../../data/dynamo", () => ({
        insertPublishedDisruptionIntoDynamoAndUpdateDraft: vi.fn(),
        getDisruptionById: vi.fn(),
        getOrganisationInfoById: vi.fn(),
    }));

    vi.mock("../../utils/apiUtils/disruptionApprovalEmailer", () => ({
        sendDisruptionApprovalEmail: vi.fn(),
    }));

    vi.mock("crypto", () => ({
        randomUUID: () => "id",
    }));

    MockDate.set("2023-03-03");

    const insertDisruptionSpy = vi.spyOn(dynamo, "insertPublishedDisruptionIntoDynamoAndUpdateDraft");
    const getDisruptionSpy = vi.spyOn(dynamo, "getDisruptionById");
    const publishSocialMediaSpy = vi.spyOn(apiUtils, "publishSocialMedia");
    const getOrganisationInfoByIdSpy = vi.spyOn(dynamo, "getOrganisationInfoById");
    const getSessionSpy = vi.spyOn(session, "getSession");
    const sendDisruptionApprovalEmailSpy = vi.spyOn(disruptionApprovalEmailer, "sendDisruptionApprovalEmail");

    beforeEach(() => {
        getOrganisationInfoByIdSpy.mockResolvedValue(orgInfo);
    });

    afterEach(() => {
        vi.resetAllMocks();
    });

    afterAll(() => {
        MockDate.reset();
    });

    it("should retrieve valid data from cookies, write to dynamo and redirect", async () => {
        getDisruptionSpy.mockResolvedValue(disruptionWithConsequences);
        const { req, res } = getMockRequestAndResponse({
            body: {
                disruptionId: defaultDisruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await publish(req, res);

        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).toBeCalledTimes(1);
        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).toBeCalledWith(
            disruptionWithConsequences,
            DEFAULT_ORG_ID,
            PublishStatus.published,
            "test@example.com",
            "Disruption created and published",
            false,
            true,
        );
        expect(writeHeadMock).toBeCalledWith(302, { Location: DASHBOARD_PAGE_PATH });
    });

    it("should retrieve valid data from cookies, write to dynamo and redirect with social media", async () => {
        getDisruptionSpy.mockResolvedValue(disruptionWithConsequencesAndSocialMediaPosts);
        const { req, res } = getMockRequestAndResponse({
            body: {
                disruptionId: defaultDisruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        publishSocialMediaSpy.mockResolvedValue();

        await publish(req, res);

        expect(publishSocialMediaSpy).toHaveBeenCalledWith(
            disruptionWithConsequencesAndSocialMediaPosts.socialMediaPosts?.filter(
                (post) => post.status === SocialMediaPostStatus.pending,
            ),
            DEFAULT_ORG_ID,
            false,
            true,
        );
        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).toBeCalledTimes(1);
        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).toBeCalledWith(
            disruptionWithConsequencesAndSocialMediaPosts,
            DEFAULT_ORG_ID,
            PublishStatus.published,
            "test@example.com",
            "Disruption created and published",
            false,
            true,
        );
        expect(writeHeadMock).toBeCalledWith(302, { Location: DASHBOARD_PAGE_PATH });
    });

    it("should redirect to error page if disruptionId not passed", async () => {
        getDisruptionSpy.mockResolvedValue(disruptionWithConsequencesAndSocialMediaPosts);
        const { req, res } = getMockRequestAndResponse({
            mockWriteHeadFn: writeHeadMock,
        });

        await publish(req, res);

        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).not.toBeCalled();
        expect(writeHeadMock).toBeCalledWith(302, { Location: ERROR_PATH });
    });

    it("should redirect to error page if disruption is invalid", async () => {
        getDisruptionSpy.mockResolvedValue({} as FullDisruption);
        const { req, res } = getMockRequestAndResponse({
            body: {
                disruptionId: defaultDisruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await publish(req, res);

        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).not.toBeCalled();
        expect(writeHeadMock).toBeCalledWith(302, { Location: ERROR_PATH });
    });

    it("should write the correct disruptions data to dynamoDB", async () => {
        getDisruptionSpy.mockResolvedValue(disruptionWithConsequencesAndSocialMediaPosts);
        const { req, res } = getMockRequestAndResponse({
            body: {
                disruptionId: disruptionWithConsequencesAndSocialMediaPosts.disruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await publish(req, res);

        expect(insertDisruptionSpy.mock.calls[0]).toMatchSnapshot();
    });

    it("should redirect to review page if no consequences", async () => {
        getDisruptionSpy.mockResolvedValue(disruptionWithNoConsequences);
        const { req, res } = getMockRequestAndResponse({
            body: {
                disruptionId: defaultDisruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await publish(req, res);

        expect(dynamo.insertPublishedDisruptionIntoDynamoAndUpdateDraft).not.toBeCalled();
        expect(writeHeadMock).toBeCalledWith(302, {
            Location: `${REVIEW_DISRUPTION_PAGE_PATH}/${defaultDisruptionId}`,
        });
    });

    it("should call sendDisruptionApprovalEmail method when an org staff creates a disruption", async () => {
        getDisruptionSpy.mockResolvedValue(disruptionWithConsequences);
        getSessionSpy.mockImplementation(() => ({ ...mockSession, isOrgStaff: true, isSystemAdmin: false }));
        const { req, res } = getMockRequestAndResponse({
            body: {
                disruptionId: defaultDisruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await publish(req, res);

        expect(sendDisruptionApprovalEmailSpy).toBeCalledWith(
            mockSession.orgId,
            disruptionWithConsequences.summary,
            disruptionWithConsequences.description,
            mockSession.name,
            disruptionWithConsequences.disruptionId,
        );
        expect(writeHeadMock).toBeCalledWith(302, {
            Location: DASHBOARD_PAGE_PATH,
        });
    });
});
