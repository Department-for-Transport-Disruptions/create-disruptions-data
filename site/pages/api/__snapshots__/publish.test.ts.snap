// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`publish > should write the correct disruptions data to dynamoDB 1`] = `
{
  "Consequences": {
    "Consequence": [
      {
        "Advice": {
          "Details": "Some consequence description",
        },
        "Affects": {
          "Networks": {
            "AffectedNetwork": {
              "AllLines": "",
              "VehicleMode": "tram",
            },
          },
        },
        "Blocking": {
          "JourneyPlanner": false,
        },
        "Condition": "unknown",
        "Severity": "slight",
      },
      {
        "Advice": {
          "Details": "Some consequence description",
        },
        "Affects": {
          "Networks": {
            "AffectedNetwork": {
              "AllLines": "",
              "VehicleMode": "bus",
            },
          },
          "Operators": {
            "AffectedOperator": [
              {
                "OperatorRef": "FSYO",
              },
            ],
          },
        },
        "Blocking": {
          "JourneyPlanner": true,
        },
        "Condition": "unknown",
        "Delays": {
          "Delay": "PT40M",
        },
        "Severity": "severe",
      },
    ],
  },
  "CreationTime": "2023-03-03T00:00:00.000Z",
  "Description": "Test description",
  "ParticipantRef": "DepartmentForTransport",
  "PersonnelReason": "staffInWrongPlace",
  "Planned": true,
  "Progress": "open",
  "PublicationWindow": {
    "StartTime": "2023-03-10T12:00:00.000Z",
  },
  "ReasonType": "PersonnelReason",
  "SituationNumber": "acde070d-8c4c-4f0d-9d8a-162843c10333",
  "Source": {
    "SourceType": "feed",
    "TimeOfCommunication": "2023-03-03T00:00:00.000Z",
  },
  "Summary": "Some summary",
  "ValidityPeriod": [
    {
      "EndTime": "2023-03-17T17:00:00.000Z",
      "StartTime": "2023-03-10T12:00:00.000Z",
    },
    {
      "StartTime": "2023-03-18T12:00:00.000Z",
    },
  ],
}
`;
